Radius

HelenMamontova opened the pull request:
Stop stargazer by ctrl+c implemented.

HelenMamontova committed:
Radius. Header file <mutex>, <jthread.hpp> added. Functions start, stop
declaration changed. Class members m_thread, m_mutex added. Function Run
declaration added.

HelenMamontova committed:
Radius. Header files added. Functions Start, Stop, Run definition added.

HelenMamontova committed:
Radius. New files server.h, server.cpp added.

HelenMamontova committed:
Radius. Definition of functions startReceive, makeResponse, handleSend,
handleReceive added.

HelenMamontova committed:
The Run function name fixed. Definition of secret variable sdded.

HelenMamontova committed:
The file server.cpp added to add_library command, command
find_package(OpenSSL 1.0.0 Required) added, OpenSSL::Crypto added to
target_link_libraries command in the block if(BUILD_MOD_RADIUS).

HelenMamontova committed:
Parameter token added to function Run() declaration.

HelenMamontova committed:
Radius. Method Start: variable isRunning=true removed, m_thread
definition changed. Method Stop: variable isRunning=false removed,
m_thread.joinable check added, isRunning check added, request_stop call added.
Method Run: parameter token added, variables secret, port removed,
object lock and isRunning=true added before cycle while,
cycle while added, isRunning=false added after cycle while.

HelenMamontova committed:
Radius. Hold the mutex removed,extra conditions
for m_thread.join() removed in the Stop function.

HelenMamontova committed:
Radius. Cycle while removed in function Run.

HelenMamontova committed:
Radius. The variables isRunning, errorStr replaced by m_running,
m_errorStr in class Radius.

HelenMamontova committed:
Radius. The variable isRunning replaced by m_running in function Run.

HelenMamontova committed:
Radius. Object name io_service changed to ioService in the function Run.

HelenMamontova committed:
Method SetRunning declaration added to the class RADIUS.

HelenMamontova committed:
Method SetRunning definition added. Method SetRunning call added to
function Run.

HelenMamontova committed:
Radius. Namespace STG added.

HelenMamontova committed:
Radius. Declaration using STG::Server added.

HelenMamontova committed:
Radius. Header file "stg/logger.h" added. RADIUS class member m_loger added.

HelenMamontova committed:
Radius. Header file "stg/common.h" added. Initialization of m_logger
added to constructor RADIUS. Output cerr replaced by logger and
printfd().

HelenMamontova committed:
Radius. Variable except added, messages fixed in the function Run.

HelenMamontova committed:
Radius. Thread join logic fixed in the function Stop.

HelenMamontova committed:
Keyword const added to std::lock_guard in the function SetRunning.

HelenMamontova committed:
Radius. Class member m_mutex moved to the top of list.

HelenMamontova committed:
Radius. Class member m_logger put after the m_running.

HelenMamontova committed:
Radius. Unnecessary variable except removed in the function Run.

HelenMamontova committed:
Radius. Extra whitespace removed in the function Stop. Function
IsRunning definition added.

HelenMamontova committed:
Radius. Function IsRunning declaration changed.

HelenMamontova committed:
Radius. Formatting fixed.

HelenMamontova committed:
Radius. Extra symbols '//' removed.

HelenMamontova committed:
Radius. Class member m_running initialization added to constructor
RADIUS.

HelenMamontova committed:
Radius.Header file "stg/module_settings.h" added. Class members variable
m_settings and method SetSettings added to class RADIUS.

HelenMamontova committed:
Radius. The parameters Secret, Dictionaries added, parameter Port
changed, other parameters removed.

HelenMamontova committed:
Merge remote-tracking branch 'madf/master' into radius

HelenMamontova committed:
Radius. Header files added. Class RAD_SETTINGS added. Methods
SetSettings, copy constructor RADIUS, assignment operator added. Method
ParseSettings changed. Formatting fixed.

HelenMamontova committed:
Radius. Parameter default changed to optional and parameters value
commented out for the parameters dictionaries, port.

HelenMamontova committed:
Radius. Description of parameters secret, port, dictionaries changed.

HelenMamontova committed:
Radius. Copy constructor RADIUS and assignment operator declaration
changed and moved to public section.

HelenMamontova committed:
Radius. Class member m_mutex moved to the top of section private.

HelenMamontova committed:
Radius. The m_port variable value initialization changed to 1812,
m_dictionaries variable initialization added to constructor RAD_SETTINGS.

HelenMamontova committed:
Radius. Check for missing parameters port and dictionaries removed in function ParseSettings.

HelenMamontova committed:
Merge remote-tracking branch 'madf/master' into radius

HelenMamontova committed:
Merge remote-tracking branch 'madf/master' into radius

HelenMamontova committed:
Radius. The dist directory added.

HelenMamontova committed:
Radius. The methods declaration stop and start added.

HelenMamontova committed:
Radius. The methods stop, start definition added. Method start call
added to Server constructor.

HelenMamontova committed:
Radius. The header files "server.h", <boost/asio.hpp>, <memory> added.
The class members m_ioService, m_server added.

HelenMamontova committed:
Radius. The header files "server", <boost/asio.hpp>, <string> removed.
The method m_server->stop() call added to method Stop.

madf commented on projects/stargazer/plugins/other/radius/radius.h:
```diff
@@ -72,5 +76,7 @@ namespace STG
 
             PluginLogger m_logger;
+
+            std::unique_ptr<Server> m_server{};
```
You do not need an initializer here.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -113,6 +88,7 @@ int RADIUS::Stop()
     if (!m_thread.joinable())
         return 0;
 
+    m_server->stop();
```
This should go after the `m_thread.request_stop()`. I try to explain why.

First, you do not use `stop_token`. You take it in lambda when you create a thread, and you pass it to the `Run`. And it is not used there in any way. It means that `m_thread.request_stop()` has no effect now. Your token turns into a 'stop requested' state, but you don't check it.

Second, when you ask the `Server` to stop, you simply close the socket. And the receiving operation that is currently waiting for data is ended with an error ('operation aborted' or something). This is not terrible, yet it is not great. When a user stops the server, the last thing he expects is to receive an obscure 'operation aborted' error in the log.

Third, your `Server` is polluted with `std::cout`. Please don't use it here. For debug purposes use `printfd` and for user interaction use the logger (`m_logger` in your code).

Taking all that into account, here is what I think. Communication errors can happen, and we should report them. Unless the errors are made by us to interrupt receiving data. When the `Server::handleReceive` is called, we need to see if we are stopping first. Because if we are, the errors do not matter and we can simply return. And if we are not, we should check for errors and report them, if there are any. The only way to see if we are stopping is to check for the stop token. There is no stop token in the `Server`, but imagine that you have it there. And you check it. And if you call the `Server` to stop before requesting the thread to stop, the stop token would not report that we are stopping.

To sum the thing up:
 - you need to reorder `m_server->stop()` and `m_thread.request_stop()`;
 - you need to check that `m_server` contains a value before calling `stop()`;
 - you need to have a stop token in the `Server`;
 - you need to check the token in handlers before making any further steps;
 - you need to replace all `std::cout` in the `Server` with either `printfd` or `m_logger` or both.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -130,3 +106,24 @@ void RADIUS::SetRunning(bool val)
+    try
+    {
+        if (!m_server)
+           m_server = std::make_unique<Server> (m_ioService, m_radSettings.GetSecret(), m_radSettings.GetPort(), m_radSettings.GetDictionaries());
```
Extra whitespace after the `>` and before the`(`.

HelenMamontova committed:
Radius. Header files <stop_token> and <jthread.hpp> added. Parameter
token added to constructor Server. Class member m_token added.

HelenMamontova committed:
Radius. Parameter token and m_token class member initialization added to Server
constructor definition. The check m_token.stop_requested() added to
method handleSend before startReceive method call.

HelenMamontova committed:
Radius. The m_server class member initialization removed.

HelenMamontova committed:
Radius. The m_server->stop() call adnd m_thread.request_stop() call
reodered, the m_server check added before m_server->stop call in the
method Stop. Parameter token added to constructor Server in the function
Run.

HelenMamontova committed:
Radius. Header file "stg/logger.h" added. Class member m_logger added to
class Server.

HelenMamontova committed:
Radius. Header file "stg/common.h" added, header file <iostream>
removed. Class member m_logger initialization added to constructor
Server. Object std::cout replaced by function printfd and m_logger.

madf commented on projects/stargazer/plugins/other/radius/server.cpp:
```diff
@@ -1,14 +1,16 @@
-      m_dictionaries(filePath)
+      m_dictionaries(filePath),
+      m_token(token),
+      m_logger(PluginLogger::get("radius"))
```
It's better to take it from the `RADIUS`.

madf commented on projects/stargazer/plugins/other/radius/server.cpp:
```diff
@@ -54,22 +56,28 @@ RadProto::Packet Server::makeResponse(const RadProto::Packet& request)
+    }
 
-    startReceive();
+    if (!m_token.stop_requested())
```
I think you should check the token before checking the error code. When you close the socket and there is a pending send or receive operation, it will result in error. However, when we are stopping, this error is expected and is not something we want to report in the log.

madf commented on projects/stargazer/plugins/other/radius/server.cpp:
```diff
@@ -54,22 +56,28 @@ RadProto::Packet Server::makeResponse(const RadProto::Packet& request)
 
 void Server::handleReceive(const error_code& error, const std::optional<RadProto::Packet>& packet, const boost::asio::ip::udp::endpoint& source)
 {
     if (error)
```
You should check for stop token here as well.

madf commented on projects/stargazer/plugins/other/radius/server.h:
```diff
@@ -3,7 +3,10 @@
+#include "stg/logger.h"
 #include <boost/asio.hpp>
+#include <stop_token>
+#include <jthread.hpp>
```
Why do you need `jthread.hpp` here?

madf commented on projects/stargazer/plugins/other/radius/server.h:
```diff
@@ -3,7 +3,10 @@
 #include "radproto/dictionaries.h"
+#include "stg/logger.h"
 #include <boost/asio.hpp>
+#include <stop_token>
```
It should be `stop_token.hpp`. We use a 3rd-party library for `std::jthread` and `std::stop_token` because compilers may not have them yet.

HelenMamontova committed:
Radius. Class member m_logger initialization changed in the constructor
Server. Check token moved before checking the error code in the
handleSend method.

HelenMamontova committed:
Radius. Check token added to handleReceive method.

HelenMamontova committed:
Radius. Header file <stop_token> replaced by <stop_token.hpp>, header
file <jthread.hpp> removed.

madf commented on projects/stargazer/plugins/other/radius/server.cpp:
```diff
@@ -10,7 +10,7 @@ Server::Server(boost::asio::io_service& io_service, const std::string& secret, u
       m_dictionaries(filePath),
       m_token(token),
-      m_logger(PluginLogger::get("radius"))
+      m_logger(PluginLogger::get("RADIUS"))
```
As I said before, you should use logger from the `RADIUS`, not get a new one from `PluginLogger::get`. Pass it from the parent by ref and store by ref.

madf commented on projects/stargazer/plugins/other/radius/server.cpp:
```diff
@@ -55,33 +55,36 @@ RadProto::Packet Server::makeResponse(const RadProto::Packet& request)
 void Server::handleSend(const error_code& ec)
 {
+    if (!m_token.stop_requested())
+        startReceive();
```
No.

1. Check if stop is requested. If it is - just return.
2. Check error code.
3. Start receive.

madf replied:
What about this?

madf commented on projects/stargazer/plugins/other/radius/server.cpp:
```diff
@@ -55,33 +55,36 @@ RadProto::Packet Server::makeResponse(const RadProto::Packet& request)
 void Server::handleReceive(const error_code& error, const std::optional<RadProto::Packet>& packet, const boost::asio::ip::udp::endpoint& source)
 {
-    if (error)
+    if (!m_token.stop_requested())
```
Same as in `handleSend`.

madf replied:
And what about this?

HelenMamontova committed:
Radius. Parameter m_logger added to constructor Server call in the
method Run.

HelenMamontova committed:
Radius. Parameter logger added to constructor Server declaration. Class
member m_logger added to class Server.

HelenMamontova committed:
Radius. Parameter logger added to constructor Server definition. Class
member m_logger initialization changed in the constructor Server.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -130,3 +108,24 @@ void RADIUS::SetRunning(bool val)
+    try
+    {
+        if (!m_server)
+           m_server = std::make_unique<Server> (m_ioService, m_radSettings.GetSecret(), m_radSettings.GetPort(), m_radSettings.GetDictionaries(), token, m_logger);
```
Extra whitespace between '>' and '('.

HelenMamontova committed:
Radius. Extra whitespace removed ib the method Run.

HelenMamontova committed:
Radius. The order of operations changed in the method handleSend.

HelenMamontova committed:
Radius. The order of operations changed in the method handleReceive.

madf commented on projects/stargazer/plugins/other/radius/server.cpp:
```diff
@@ -55,36 +55,35 @@ RadProto::Packet Server::makeResponse(const RadProto::Packet& request)
-        }
+        m_logger("Error asyncReceive: %s", error.message());
+        printfd(__FILE__, "Error asyncReceive: '%s'\n", error.message());
+        return;
```
It's interesting. In `handleSend` on error you log it and proceed. In `handleReceive` on error you log it and stop the server. Error handling should be uniform, you either proceed or stop. And since this is a server and it should be robust, it is better to proceed. Please make changes accordingly.

madf commented on projects/stargazer/plugins/other/radius/server.cpp:
```diff
@@ -55,36 +55,35 @@ RadProto::Packet Server::makeResponse(const RadProto::Packet& request)
+        printfd(__FILE__, "Error asyncReceive: the request packet is missing\n");
+        return;
     }
+    else
```
You do not need `else` if the previous branch has a `return`.

HelenMamontova committed:
Radius. Operator return removed when checking ec, operator else removed
when checking packet in the handleReceive method.

madf commented on projects/stargazer/plugins/other/radius/radius.cpp:
```diff
@@ -130,3 +108,24 @@ void RADIUS::SetRunning(bool val)
+    try
+    {
+        if (!m_server)
+           m_server = std::make_unique<Server>(m_ioService, m_radSettings.GetSecret(), m_radSettings.GetPort(), m_radSettings.GetDictionaries(), token, m_logger);
```
`std::move(token)`

madf commented on projects/stargazer/plugins/other/radius/server.cpp:
```diff
@@ -1,18 +1,31 @@
     : m_radius(io_service, secret, port),
-      m_dictionaries(filePath)
+      m_dictionaries(filePath),
+      m_token(token),
```
`std::move(token)`

madf commented on projects/stargazer/plugins/other/radius/server.cpp:
```diff
@@ -42,27 +55,33 @@ RadProto::Packet Server::makeResponse(const RadProto::Packet& request)
+    if (ec)
+    {
+        m_logger("Error asyncSend: %s", ec.message());
+        printfd(__FILE__, "Error asyncSend: '%s'\n", ec.message());
```
`ec.message()` returns `std::string`, while '%s' requires `const char*`. Please add `.c_str()` here and in other places.

HelenMamontova committed:
Radius. Parameter token replaced by std::move(token) in the constructor Server
in method Run.

HelenMamontova committed:
Radius. Class member m_token initialization value changed to
std::move(token) in the constructor Server.

HelenMamontova committed:
Radius. The c_str() function added to message when checking error in the
handleSend and handleReceive methods.

HelenMamontova committed:
Radius. Header file <cstring> added.
